ern-core/src/FsCache.ts

```diff
@@ -0,0 +1,220 @@
+    }
+    if (!objectToId) {
+      throw new Error('objectToId function is required')
+    }
+    if (!rootCachePath) {
+      throw new Error('rootCachePath is required')
+    }
+    this.rootCachePath = rootCachePath
+    this.maxCacheSize = maxCacheSize
+    this.addObjectToDirectory = addObjectToDirectory
+    this.objectToId = objectToId
+    if (!fs.existsSync(rootCachePath)) {
+      shell.mkdir(rootCachePath)
+    }
+    if (!fs.existsSync(this.cacheManifestPath)) {
+      writeJSONSync(this.cacheManifestPath, { entries: [], size: 0 })
+    }
+  }
+
+  /**
+   * Indicates whether a given object is currently stored in the cache
+   * @param obj The object to check for presence in the cache
+   */
+  public async isInCache(obj: T): Promise<boolean> {
+    const manifestObj = await this.getManifestObj()
+    return !!manifestObj.entries.find(e => e.id === this.objectToId(obj))
+  }
+
+  /**
+   * Adds an object entry to the cache
+   * @param obj The object to add to the cache
+   */
+  public async addToCache(obj: T): Promise<string> {
+    if (await this.isInCache(obj)) {
+      throw new Error(
+        `Object with id ${this.objectToId(obj)} is already in the cache`
+      )
+    }
+    // Create unique directory to store the object
+    const pathToObjectCacheDir = this.createUniqueCacheDirectory()
+    // Invoke client provided function to write object in cache directory
+    await this.addObjectToDirectory(obj, pathToObjectCacheDir)
+    // Compute cached object directory size
+    const cachedObjectDirSize = await getSizeAsync(pathToObjectCacheDir)
+    // Create manifest entry for new cached object
+    const entry: CacheManifestEntry = {
+      id: this.objectToId(obj),
+      lastAccessed: Date.now(),
+      path: pathToObjectCacheDir,
+      size: cachedObjectDirSize,
+    }
+    // Update manifest total cache size
+    const manifestObj = await this.getManifestObj()
+    manifestObj.size += cachedObjectDirSize
+    manifestObj.entries.push(entry)
+    // Clean cache if newly added object caused total cache
+    // size to exceed the defined max cache size
+    // Remove all entries that were least recently accessed
+    // until the total cache size falls below the max cache size
+    const entriesSortedByLat = manifestObj.entries.sort(
+      (a, b) => b.lastAccessed - a.lastAccessed
+    )
+    while (manifestObj.size >= this.maxCacheSize) {
+      const entryToRemove = entriesSortedByLat.pop()
+      manifestObj.size -= entryToRemove!.size
+      shell.rm('-rf', entryToRemove!.id)
+    }
+    manifestObj.entries = entriesSortedByLat
+    // Save updated manifest
+    await writeJSON(this.cacheManifestPath, manifestObj)
+    return pathToObjectCacheDir
+  }
+
+  /**
+   * Gets path to the directory containing a cached object
+   * @param obj The object for which to retrieve cached directory path
+   */
+  public async getObjectCachePath(obj: T): Promise<string | void> {
+    const manifestObj = await this.getManifestObj()
```

Comment:
If it's possible to read `return readJSONSync(this.cacheManifestPath)`, we can `await writeJSONSync(this.cacheManifestPath, manifestObj)` (line 193) ? In which case we can avoid `await`

Reply:
We should use asynchronous methods as much as possible in the code base. 
Async (promise based) is the way to go everywhere doable.
Unfortunately, constructors cannot be executed asynchronously, thus there was no other choice but to choose sync version in the constructor (unless change of design with lazy init, but wanted to keep things simple). That's not a big issue.